semantic-core/src/Data/Scope.hs

```diff
@@ -0,0 +1,102 @@
+{-# LANGUAGE DeriveTraversable, LambdaCase, RankNTypes, QuantifiedConstraints, StandaloneDeriving #-}
+module Data.Scope
+( Incr(..)
+, incr
+, Scope(..)
+, fromScope
+, toScope
+, bind1
+, bind
+, bindEither
+, instantiate1
+, instantiate
+, instantiateEither
+) where
+
+import Control.Applicative (liftA2)
+import Control.Effect.Carrier
+import Control.Monad ((>=>), guard)
+import Control.Monad.Module
+import Control.Monad.Trans.Class
+import Data.Function (on)
+
+data Incr a b
+  = Z a
+  | S b
+  deriving (Eq, Foldable, Functor, Ord, Show, Traversable)
+
+instance Applicative (Incr a) where
+  pure = S
+  Z a <*> _ = Z a
+  S f <*> a = f <$> a
+
+instance Monad (Incr a) where
+  Z a >>= _ = Z a
+  S a >>= f = f a
+
+match :: Applicative f => (b -> Either a c) -> b -> Incr a (f c)
+match f x = either Z (S . pure) (f x)
+
+matchMaybe :: (b -> Maybe a) -> (b -> Either a b)
+matchMaybe f a = maybe (Right a) Left (f a)
+
+incr :: (a -> c) -> (b -> c) -> Incr a b -> c
+incr z s = \case { Z a -> z a ; S b -> s b }
+
+
+newtype Scope a f b = Scope { unScope :: f (Incr a (f b)) }
```

Comment:
`Scope` embeds an `f` with a bound variable, represented as `f` wrapping `Incr` wrapping `f`.

1. The outer `f` is to embed subterms, necessary for any reasonable binding syntax.

2. `Incr` is used to ensure that the type of the embedded term differs from the type of the embedding term; “nested datatypes” of this nature enable us to use the type system to ensure that we aren’t unwrapping scopes without correctly handling their bound variables, which is key for ensuring the correctness of finicky operations like capture-avoiding substitution, calculating free variables, and so on.

3. The inner `f` enables us to instantiate a `Scope`’s bound variable with an entire argument _without_ changing that argument; in effect, instead of pushing the `Incr`’s `S` constructors all the way through the argument to wrap each new bound variable, we wrap entire subterms with `S` instead. This is more efficient, tho it also makes certain folds kind of painful.

Using this inside syntactic binders like `Lam` is what gives us scope safety!